File: ios/Classes/IvsPlayerView.swift
import UIKit
import AVFoundation
import Flutter
import AmazonIVSPlayer

class IvsPlayerView: NSObject, FlutterPlatformView, FlutterStreamHandler , IVSPlayer.Delegate{
    
    private var playerView: UIView
    private var _methodChannel: FlutterMethodChannel?
    private var _eventChannel: FlutterEventChannel?
    private var _eventSink: FlutterEventSink?
    private var players: [String: IVSPlayer] = [:] // Dictionary to manage multiple players
    private var playerViews: [String: IVSPlayerView] = [:]
    private var playerId: String?
    
    func view() -> UIView {
        return playerView;
    }
    
    func player(_ player: IVSPlayer, didChangeState state: IVSPlayer.State) {
        guard let eventSink = _eventSink else { return }
        let dict: [String: Any] = ["state": state.rawValue]
        print(dict)
        eventSink(dict)
    }

    func player(_ player: IVSPlayer, didChangeDuration time: CMTime) {
        guard let eventSink = _eventSink else { return }
        let dict: [String: Any] = ["duration": time.seconds]
        print(dict)
        eventSink(dict)
    }

    func player(_ player: IVSPlayer, didChangeSyncTime time: CMTime) {
        guard let eventSink = _eventSink else { return }
        let dict: [String: Any] = ["syncTime": time.seconds]
        print(dict)
        eventSink(dict)
    }

    func player(_ player: IVSPlayer, didChangeQuality quality: IVSQuality?) {
        guard let eventSink = _eventSink else { return }
        let dict: [String: Any] = ["quality": quality?.name ?? ""]
        print(dict)
        eventSink(dict)
    }
    
    func player(_ player: IVSPlayer, didOutputCue cue: IVSCue) {
        if let textMetadataCue = cue as? IVSTextMetadataCue {
            let dict: [String: Any] = [
                "metadata": textMetadataCue.text,
                "startTime": textMetadataCue.startTime.epoch,
                "endTime": textMetadataCue.endTime.epoch,
            ]
            _eventSink?(dict)
        }
    }

    func player(_ player: IVSPlayer, didFailWithError error: any Error) {
        guard let eventSink = _eventSink else { return }
        let dict: [String: Any] = ["error": error.localizedDescription]
        print(dict)
        eventSink(dict)
    }

    func player(_ player: IVSPlayer, didSeekTo time: CMTime) {
        guard let eventSink = _eventSink else { return }
        let dict: [String: Any] = ["seekedToTime": time.seconds]
        print(dict)
        eventSink(dict)
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self._eventSink = events;
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self._eventSink = nil
        return nil
    }
    
    init(_ frame: CGRect,
         viewId: Int64,
         args: Any?,
         messenger: FlutterBinaryMessenger
    ) {
        _methodChannel = FlutterMethodChannel(
            name: "ivs_player", binaryMessenger: messenger
        );
        _eventChannel = FlutterEventChannel(name: "ivs_player_event", binaryMessenger: messenger)
        playerView = UIView(frame: frame)
        super.init();
        _methodChannel?.setMethodCallHandler(onMethodCall)
        _eventChannel?.setStreamHandler(self)
    }
    
    
    func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
        print(call.method)
        switch(call.method) {
        case "createPlayer":
            let args = call.arguments as? [String: Any]
            let playerId = args?["playerId"] as? String
            createPlayer(playerId: playerId!)
            result(true)
        case "multiPlayer":
            let args = call.arguments as? [String: Any]
            let urls = args?["urls"] as? [String]
            multiPlayer(urls!)
            result("Players created successfully")
        case "selectPlayer":
            let args = call.arguments as? [String: Any]
            let playerId = args?["playerId"] as? String
            selectPlayer(playerId: playerId!)
            result(true)
        case "startPlayer":
            let args = call.arguments as? [String: Any]
            let url = args?["url"] as? String
            let autoPlay = args?["autoPlay"] as? Bool
            startPlayer(  url: url!, autoPlay: autoPlay!)
            result(true)
        case "stopPlayer":
            let playerId = self.playerId
            if (playerId != nil){
                stopPlayer(playerId: playerId!)
            }
            result(true)
        case "mute":
            let playerId = self.playerId
            mutePlayer(playerId: playerId!)
            result(true)
        case "pause":
            let playerId = self.playerId
            pausePlayer(playerId: playerId!)
            result(true)
        case "resume":
            let playerId = self.playerId
            resumePlayer(playerId: playerId!)
            result(true)
        case "seek":
            let args = call.arguments as? [String: Any]
            let playerId = self.playerId
            let time = args?["time"] as? String
            seekPlayer(playerId: playerId!, time!)
            result(true)
        case "position":
            let playerId = self.playerId
            result(getPosition(playerId: playerId!))
        case "qualities":
            let playerId = self.playerId
            if(playerId != nil){
                let qualities = getQualities(playerId: playerId!)
                result(qualities)
            }else {
                result([])
            }
        case "setQuality":
            let args = call.arguments as? [String: Any]
            let playerId = self.playerId
            let quality = args?["quality"] as? String
            setQuality(playerId: playerId!, quality!)
            result(true)
        case "autoQuality":
            let playerId = self.playerId
            toggleAutoQuality(playerId: playerId!)
            result(true)
        case "isAuto":
            let playerId = self.playerId
            
            result(isAuto(playerId: playerId!))
        case "getScreenshot":
            let args = call.arguments as? [String: Any]
            let url = args?["url"] as? String
            result(getScreenShot(url: url!))
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    
    
    
    func getScreenShot(url: String) -> [UInt8]? {
        guard let videoURL = URL(string: url) else {
            print("Invalid URL")
            return nil
        }
        
        // Create an AVAsset and AVAssetImageGenerator
        let asset = AVAsset(url: URL(string: "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4")!)
        let imageGenerator = AVAssetImageGenerator(asset: asset)
        imageGenerator.appliesPreferredTrackTransform = true // Adjust for orientation
        
        // Define the time for the screenshot (e.g., at the 1-second mark)
        let time = CMTime(seconds: 1.0, preferredTimescale: 600)
        
        do {
            // Generate the CGImage
            let cgImage = try imageGenerator.copyCGImage(at: time, actualTime: nil)
            // Convert to UIImage
            let image =  UIImage(cgImage: cgImage)
            guard let imageData = image.pngData() else { return nil }
            return [UInt8](imageData)
        } catch {
            print("Failed to generate screenshot: \(error.localizedDescription)")
            return nil
        }
        
    }
    
    func multiPlayer(_ urls:[String]) {
        self.playerId = urls.first!
        for url in urls {
            let player = IVSPlayer()
            player.delegate = self
            let playerId = url
            players[playerId] = player
            playerViews[playerId] = IVSPlayerView()
            playerViews[playerId]?.player = player
            player.load(URL(string: url))
            player.volume = 0
        }
        // Play All Players and attach to the first player to preview
        for (index, player) in players.enumerated() {
            if index == 0 {
                player.value.play()
                player.value.volume = 1
                attachPreview(container: self.playerView, preview: playerViews[player.key]!)
            } else {
                player.value.play()
            }
        }
    }
    
    func selectPlayer(playerId: String) {
        guard let player = players[playerId] else { return }
        let previousPlayer = playerId == self.playerId ? nil : self.playerId
        if let previousPlayer {
            players[previousPlayer]?.delegate = nil
            players[previousPlayer]?.volume = 0
        }
        self.playerId = playerId
        let playerView = playerViews[playerId]!
        player.volume = 1
        player.delegate = self
        UIView.animate(withDuration: 0.1, animations: {
            self.playerView.alpha = 0
        }) { _ in
            // Update the preview
            self.attachPreview(container: self.playerView, preview: playerView)
            
            // Fade in the new preview
            UIView.animate(withDuration: 0.5) {
                self.playerView.alpha = 1
            }
        }
        updateEventsOfCurrentPlayer()
    }
    
    func updateEventsOfCurrentPlayer() {
        guard let player = players[self.playerId!] else { return }
        player.delegate = self 
        let dict: [String: Any] = ["state": player.state.rawValue, "duration": player.duration.seconds, "syncTime": player.syncTime.seconds, "quality": player.quality?.name ?? ""]
        print(dict)
        _eventSink?(dict)
    }
    
    func stopPlayer(playerId: String) {
        guard let player = players[playerId] else { return }
        player.pause()
        players.removeValue(forKey: playerId)
        playerViews.removeValue(forKey: playerId)
    }
    
    func mutePlayer(playerId: String) {
        guard let player = players[playerId] else { return }
        player.volume = player.volume == 0 ? 1 : 0
    }
    
    func pausePlayer(playerId: String) {
        guard let player = players[playerId] else { return }
        player.pause()
    }
    
    func resumePlayer(playerId: String) {
        guard let player = players[playerId] else { return }
        player.play()
    }
    
    func seekPlayer(playerId: String, _ time: String) {
        guard let player = players[playerId] else { return }
        player.seek(to: CMTimeMake(value: Int64(time) ?? 0, timescale: 1))
    }
    
    func getPosition(playerId: String) -> String {
        guard let player = players[playerId] else { return "0" }
        return player.position.seconds.description
    }
    
    func getQualities(playerId: String) -> [String] {
        guard let player = players[playerId] else { return [] }
        return player.qualities.map { $0.name }
    }
    
    func setQuality(playerId: String, _ quality: String) {
        guard let player = players[playerId] else { return }
        let qualities = player.qualities
        let qualityToChange = qualities.first { $0.name == quality }
        player.setQuality(qualityToChange!, adaptive: true)
    }
    
    func toggleAutoQuality(playerId: String) {
        guard let player = players[playerId] else { return }
        player.autoQualityMode.toggle()
    }
    
    func isAuto(playerId: String) -> Bool {
        guard let player = players[playerId] else { return false }
        return player.autoQualityMode
    }
    
    func createPlayer(playerId: String) {
        if( players[playerId] != nil ){
            return
        }
        let player = IVSPlayer()
        player.delegate = self
        self.playerId = playerId
        players[playerId] = player
        playerViews[playerId] = IVSPlayerView()
        playerViews[playerId]?.player = player
        player.load(URL(string: playerId))
    }
    
    func startPlayer(url: String, autoPlay: Bool){
        guard let player = players[url], let _ = playerViews[url] else {
            return
        }
        self.playerId = url
        if autoPlay {
            player.play()
        }
        selectPlayer(playerId: self.playerId!)
    }
    
    func attachPreview(container: UIView, preview: UIView) {
        // Clear current view, and then attach the new view.
        container.subviews.forEach { $0.removeFromSuperview() }
        preview.translatesAutoresizingMaskIntoConstraints = false
        container.addSubview(preview)
        NSLayoutConstraint.activate([
            preview.topAnchor.constraint(equalTo: container.topAnchor, constant: 0),
            preview.bottomAnchor.constraint(equalTo: container.bottomAnchor, constant: 0),
            preview.leadingAnchor.constraint(equalTo: container.leadingAnchor, constant: 0),
            preview.trailingAnchor.constraint(equalTo: container.trailingAnchor, constant: 0),
        ])
    }
    
}

File: android/src/main/java/com/example/ivs_broadcaster/IvsPlayerView.java
package com.example.ivs_broadcaster;


import android.content.Context;
import android.net.Uri;
import android.view.Surface;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.amazonaws.ivs.player.Cue;
import com.amazonaws.ivs.player.Player;
import com.amazonaws.ivs.player.PlayerException;
import com.amazonaws.ivs.player.PlayerView;
import com.amazonaws.ivs.player.Quality;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.platform.PlatformView;

public class IvsPlayerView extends Player.Listener implements PlatformView, SurfaceHolder.Callback, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private final Player player;
    private final SurfaceView surfaceView;
    private Surface surface;
    private EventChannel.EventSink eventSink;

    public IvsPlayerView(Context context, BinaryMessenger messenger, int viewId, Object args) {
        this.player = new PlayerView(context).getPlayer();
        this.player.addListener(this);
        this.surfaceView = new SurfaceView(context);
        MethodChannel methodChannel = new MethodChannel(messenger, "ivs_player");
        methodChannel.setMethodCallHandler(this);
        this.surfaceView.getHolder().addCallback(this);
        EventChannel eventChannel = new EventChannel(messenger, "ivs_player_event");
        eventChannel.setStreamHandler(this);
    }

    @Nullable
    @Override
    public View getView() {
        return surfaceView;
    }

    private void sendEvent(Object data) {
        if (eventSink != null) {
            eventSink.success(data);
        }
    }

    @Override
    public void onFlutterViewAttached(@NonNull View flutterView) {
        PlatformView.super.onFlutterViewAttached(flutterView);
        surfaceView.getHolder().addCallback(this);
    }

    @Override
    public void onFlutterViewDetached() {
        PlatformView.super.onFlutterViewDetached();
    }

    @Override
    public void dispose() {
        player.removeListener(this);
        player.release();
    }

    @Override
    public void surfaceCreated(@NonNull SurfaceHolder holder) {
        this.surface = holder.getSurface();
        if (player != null) {
            player.setSurface(this.surface);
        }
    }

    @Override
    public void surfaceChanged(@NonNull SurfaceHolder holder, int format, int width, int height) {

    }

    @Override
    public void surfaceDestroyed(@NonNull SurfaceHolder holder) {
        this.surface = null;
        if (player != null) {
            player.setSurface(null);
        }
    }

    @Override
    public void onListen(Object arguments, EventChannel.EventSink events) {
        this.eventSink = events;

    }

    @Override
    public void onCancel(Object arguments) {
        this.eventSink = null;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {

        Map<String, Object> args = call.arguments();
        switch (call.method) {
            case "startPlayer":
                assert args != null;
                String url = (String) args.get("url");
                Boolean autoPlay = (Boolean) args.get("autoPlay");
                startPlayer(url, autoPlay);
                result.success(true);
                break;
            case "stopPlayer":
                stopPlayer();
                result.success(true);
                break;
            case "mute":
                mutePlayer();
                result.success(true);
                break;
            case "pause":
                pausePlayer();
                result.success(true);
                break;
            case "resume":
                resumePlayer();
                result.success(true);
                break;
            case "seek":
                args = call.arguments();
                String time = (String) args.get("time");
                seekPlayer(time);
                result.success(true);
                break;
            case "position":
                result.success(player.getPosition());
                break;
            case "qualities":
                List<String> qualities = getQualities();
                result.success(qualities);
                break;
            case "setQuality":
                args = call.arguments();
                String quality = (String) args.get("quality");
                setQuality(quality);
                result.success(true);
                break;
            case "autoQuality":
                toggleAutoQuality();
                result.success(true);
                break;
            case "isAuto":
                result.success(isAuto());
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void seekPlayer(String time) {
        if (player != null) {
            player.seekTo(Long.parseLong(time));
        }
    }

    private void resumePlayer() {
        if (player != null) {
            player.play();
        }
    }

    private void pausePlayer() {
        if (player != null) {
            player.pause();
        }
    }

    private void mutePlayer() {
        if (player != null) {
            player.setMuted(!player.isMuted());
        }
    }

    private void stopPlayer() {
        this.dispose();
    }

    private void startPlayer(String url, Boolean autoPlay) {
        if (player != null) {
            player.load(Uri.parse(url));
            if (autoPlay) {
                player.play();
            }
        }
    }

    private void toggleAutoQuality() {
        final boolean auto = isAuto();
        player.setAutoQualityMode(!auto);
    }

    private boolean isAuto() {
        return player.isAutoQualityMode();
    }

    private void setQuality(String quality) {
        if (player != null) {
            for (Quality q : player.getQualities()) {
                if (q.getName().equals(quality)) {
                    player.setQuality(q);
                    break;
                }
            }
        }
    }

    private ArrayList<String> getQualities() {
        ArrayList<String> qualities = new ArrayList<>();
        if (player != null) {
            for (Quality quality : player.getQualities()) {
                qualities.add(quality.getName());
            }
        }
        return qualities;
    }

    @Override
    public void onCue(@NonNull Cue cue) {

    }

    @Override
    public void onDurationChanged(long l) {
        HashMap<String, Object> data = new HashMap<>();
        data.put("duration", l);
        sendEvent(data);
    }

    @Override
    public void onStateChanged(@NonNull Player.State state) {

        HashMap<String, Object> data = new HashMap<>();
        data.put("state", state.ordinal());
        sendEvent(data);
    }

    @Override
    public void onError(@NonNull PlayerException e) {
        HashMap<String, Object> data = new HashMap<>();
        data.put("error", e.getMessage());
        sendEvent(data);
    }

    @Override
    public void onRebuffering() {

    }

    @Override
    public void onSeekCompleted(long l) {
        HashMap<String, Object> data = new HashMap<>();
        data.put("seekedtotime", l);
        sendEvent(data);
    }

    @Override
    public void onVideoSizeChanged(int i, int i1) {

    }

    @Override
    public void onQualityChanged(@NonNull Quality quality) {
        HashMap<String, Object> data = new HashMap<>();
        data.put("quality", quality.getName());
        sendEvent(data);
    }
}
